src/IO/H5/Helpers.cpp

```diff
@@ -339,24 +442,64 @@ bool contains_attribute(const hid_t file_id, const std::string& group_name,
          std::end(names);
 }
 
-DataVector read_data(const hid_t group_id, const std::string& dataset_name) {
-  // Read a DataVector from the group
-  const hid_t dataset_id =
-      H5Dopen2(group_id, dataset_name.c_str(), h5p_default());
-  CHECK_H5(dataset_id, "could not open dataset '" << dataset_name << "'");
-  // Get the number of points. These do not need a "close" call.
-  const hid_t space_id = H5Dget_space(dataset_id);
-  CHECK_H5(space_id, "Failed to open dataspace");
-  const hid_t number_of_points = H5Sget_simple_extent_npoints(space_id);
-  CHECK_H5(number_of_points, "Failed to get number of points");
-  H5Sclose(space_id);
-  // Load the data
-  DataVector data(static_cast<size_t>(number_of_points));
-  CHECK_H5(H5Dread(dataset_id, h5_type<double>(), h5s_all(), h5s_all(),
-                   h5p_default(), static_cast<void*>(data.data())),
-           "Failed to read data");
-  CHECK_H5(H5Dclose(dataset_id), "Failed to close dataset");
-  return data;
+hid_t open_dataset(const hid_t group_id,
+                   const std::string& dataset_name) noexcept {
+  const hid_t dataset_id = H5Dopen2(
+                           group_id, dataset_name.c_str(), h5p_default());
+  CHECK_H5(dataset_id, "Failed to open dataset '" << dataset_name << "'");
+  return dataset_id;
+}
+
+void close_dataset(const hid_t dataset_id) noexcept {
+CHECK_H5(H5Dclose(dataset_id), "Failed to close dataset");
+}
+
+hid_t open_dataspace(const hid_t dataset_id) noexcept {
+  const hid_t dataspace_id = H5Dget_space(dataset_id);
+  CHECK_H5(dataspace_id, "Failed to open dataspace");
+  return dataspace_id;
+}
+
+void close_dataspace(const hid_t dataspace_id) noexcept {
+  CHECK_H5(H5Sclose(dataspace_id), "Failed to close dataspace");
+}
+
+template <size_t Rank, typename T>
+T read_data(const hid_t group_id, const std::string& dataset_name) noexcept {
+  const hid_t dataset_id = open_dataset(group_id, dataset_name);
+  const hid_t dataspace_id = open_dataspace(dataset_id);
+  std::array<hsize_t, Rank == 0 ? 1 : Rank> size{}, max_size{};
+  if (static_cast<int>(Rank) !=
+      H5Sget_simple_extent_dims(dataspace_id, size.data(), max_size.data())) {
+    ERROR("Incorrect rank in get_data(). Expected rank = "
+          << Rank << " but received array with rank = "
+          << H5Sget_simple_extent_dims(dataspace_id, size.data(),
+                                       max_size.data()));
+  }
+  close_dataspace(dataspace_id);
+
+  // Load data from the H5 file by passing a pointer to 'data' to H5Dread
+  // The type of 'data' is determined by the template parameter 'T'
+  const size_t size_product = std::accumulate(
```

Comment:
Is there not a more descriptive name than `size_product`? Isn't this the number of components?

Reply:
It's the product of the number of components along each dimension (eg. if there are 3 dimensions with # of components = 2,3,4, then it returns 24). So it is the product of the 'size' of each dimension. Also it is multiplying together individual components of the array named 'size'. Do you have a suggestion for a better name?

Reply:
Maybe `total_number_of_components`?